executor/executor.go

```diff
@@ -1685,6 +1685,17 @@ func ResetContextOfStmt(ctx sessionctx.Context, s ast.StmtNode) (err error) {
 			return err
 		}
 		s = prepareStmt.PreparedAst.Stmt
+		is := ctx.GetInfoSchema()
+		if prepareStmt.PreparedAst.SchemaVersion != is.SchemaMetaVersion() {
+			prepareStmt.PreparedAst.CachedPlan = nil
+			prepareStmt.Executor = nil
+			ret := &plannercore.PreprocessorReturn{InfoSchema: is.(infoschema.InfoSchema)}
+			err := plannercore.Preprocess(ctx, s, plannercore.InPrepare, plannercore.WithPreprocessorReturn(ret))
+			if err != nil {
+				return plannercore.ErrSchemaChanged.GenWithStack("Schema change caused error: %s", err.Error())
+			}
+			prepareStmt.PreparedAst.SchemaVersion = is.SchemaMetaVersion()
+		}
```

Comment:
IIRC, `OptimizePreparedPlan` is a must executed code path in both binary mysql protocol and text protocol. There is no need to execute it here, it is just a repetition.